From 37f6fa97590f73c42bcb497c866ea73c28cd97a5 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 31 Aug 2023 15:55:36 +0200
Subject: [PATCH 36/65] rng: stm32: implement support for STM32MP25x platforms

Implement the support for STM32MP25x platforms. On this platform, a
security clock is shared between some hardware blocks. For the RNG,
it is the RNG kernel clock. Therefore, the gate is no more shared
between the RNG bus and kernel clocks as on STM32MP1x platforms and
the bus clock has to be managed on its own.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git c220a41ff2bf945ce0189d40dfc95335dba587c5 rng: stm32: implement support for STM32MP25x platforms]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I0a7dc90c481acee91bb2aab6916cc6a5303bcf01
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/325311
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/rng/stm32_rng.c | 55 ++++++++++++++++++++++++++++++++++-------
 1 file changed, 46 insertions(+), 9 deletions(-)

diff --git a/drivers/rng/stm32_rng.c b/drivers/rng/stm32_rng.c
index 44e8a460752..aff0b4915e8 100644
--- a/drivers/rng/stm32_rng.c
+++ b/drivers/rng/stm32_rng.c
@@ -48,6 +48,7 @@
  * struct stm32_rng_data - RNG compat data
  *
  * @max_clock_rate:	Max RNG clock frequency, in Hertz
+ * @nb_clock:		Number of clock to handle
  * @cr:			Entropy source configuration
  * @nscr:		Noice sources control configuration
  * @htcr:		Health tests configuration
@@ -56,6 +57,7 @@
  */
 struct stm32_rng_data {
 	uint max_clock_rate;
+	uint nb_clock;
 	u32 cr;
 	u32 nscr;
 	u32 htcr;
@@ -65,6 +67,7 @@ struct stm32_rng_data {
 struct stm32_rng_plat {
 	fdt_addr_t base;
 	struct clk clk;
+	struct clk bus_clk;
 	struct reset_ctl rst;
 	const struct stm32_rng_data *data;
 	bool ced;
@@ -258,6 +261,14 @@ static int stm32_rng_init(struct stm32_rng_plat *pdata)
 	if (err)
 		return err;
 
+	if (pdata->data->nb_clock > 1) {
+		err = clk_enable(&pdata->bus_clk);
+		if (err) {
+			clk_disable(&pdata->clk);
+			return err;
+		}
+	}
+
 	cr = readl(pdata->base + RNG_CR);
 
 	/*
@@ -320,8 +331,16 @@ static int stm32_rng_init(struct stm32_rng_plat *pdata)
 
 static int stm32_rng_cleanup(struct stm32_rng_plat *pdata)
 {
+	int err;
+
 	writel(0, pdata->base + RNG_CR);
 
+	if (pdata->data->nb_clock > 1) {
+		err = clk_disable(&pdata->bus_clk);
+		if (err)
+			return err;
+	}
+
 	return clk_disable(&pdata->clk);
 }
 
@@ -329,8 +348,6 @@ static int stm32_rng_probe(struct udevice *dev)
 {
 	struct stm32_rng_plat *pdata = dev_get_plat(dev);
 
-	pdata->data = (struct stm32_rng_data *)dev_get_driver_data(dev);
-
 	reset_assert(&pdata->rst);
 	udelay(20);
 	reset_deassert(&pdata->rst);
@@ -354,9 +371,21 @@ static int stm32_rng_of_to_plat(struct udevice *dev)
 	if (!pdata->base)
 		return -ENOMEM;
 
-	err = clk_get_by_index(dev, 0, &pdata->clk);
-	if (err)
-		return err;
+	pdata->data = (struct stm32_rng_data *)dev_get_driver_data(dev);
+
+	if (pdata->data->nb_clock > 1) {
+		err = clk_get_by_name(dev, "rng_clk", &pdata->clk);
+		if (err)
+			return err;
+
+		err = clk_get_by_name(dev, "rng_hclk", &pdata->bus_clk);
+		if (err)
+			return err;
+	} else {
+		err = clk_get_by_index(dev, 0, &pdata->clk);
+		if (err)
+			return err;
+	}
 
 	err = reset_get_by_index(dev, 0, &pdata->rst);
 	if (err)
@@ -371,9 +400,19 @@ static const struct dm_rng_ops stm32_rng_ops = {
 	.read = stm32_rng_read,
 };
 
+static const struct stm32_rng_data stm32mp25_rng_data = {
+	.has_cond_reset = true,
+	.max_clock_rate = 48000000,
+	.nb_clock = 2,
+	.htcr = 0x969D,
+	.nscr = 0x2B5BB,
+	.cr = 0xF00D00,
+};
+
 static const struct stm32_rng_data stm32mp13_rng_data = {
 	.has_cond_reset = true,
 	.max_clock_rate = 48000000,
+	.nb_clock = 1,
 	.htcr = 0x969D,
 	.nscr = 0x2B5BB,
 	.cr = 0xF00D00,
@@ -382,13 +421,11 @@ static const struct stm32_rng_data stm32mp13_rng_data = {
 static const struct stm32_rng_data stm32_rng_data = {
 	.has_cond_reset = false,
 	.max_clock_rate = 3000000,
-	/* Not supported */
-	.htcr = 0,
-	.nscr = 0,
-	.cr = 0,
+	.nb_clock = 1,
 };
 
 static const struct udevice_id stm32_rng_match[] = {
+	{.compatible = "st,stm32mp25-rng", .data = (ulong)&stm32mp25_rng_data},
 	{.compatible = "st,stm32mp13-rng", .data = (ulong)&stm32mp13_rng_data},
 	{.compatible = "st,stm32-rng", .data = (ulong)&stm32_rng_data},
 	{},
-- 
2.51.0

